src/hotspot/share/opto/divnode.cpp

```diff
@@ -1084,12 +1149,97 @@ const Type* ModINode::Value(PhaseGVN* phase) const {
 
 //=============================================================================
 //------------------------------Idealize---------------------------------------
-Node *UModINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+
+template <typename TypeClass, typename Signed>
+static Node* unsigned_mod_ideal(PhaseGVN* phase, bool can_reshape, Node* mod) {
   // Check for dead control input
-  if( in(0) && remove_dead_region(phase, can_reshape) )  return this;
+  if (mod->in(0) && mod->remove_dead_region(phase, can_reshape)) {
+    return mod;
+  }
+  // Don't bother trying to transform a dead node
+  if (mod->in(0) && mod->in(0)->is_top()) {
+    return nullptr;
+  }
+
+  // Get the modulus
+  const Type* t = phase->type(mod->in(2));
+  if (t == Type::TOP) {
+    return nullptr;
+  }
+  const TypeClass* ti = t->is<TypeClass>();
+
+  // Check for useless control input
+  // Check for excluding mod-zero case
+  if (mod->in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    mod->set_req(0, nullptr); // Yank control input
+    return mod;
+  }
+
+  if (!ti->is_con()) {
+    return nullptr;
+  }
+  Signed con = ti->get_con();
+
+  if (con == 1) {
+    return ConNode::make(TypeClass::ZERO);
+  }
+  if (con == 0) {
+    return nullptr;
+  }
+
+  if (is_power_of_2(con)) {
+    return make_and<TypeClass>(mod->in(1), phase->makecon(TypeClass::make(con - 1)));
+  }
+
   return nullptr;
 }
 
+template <typename TypeClass, typename Unsigned, typename Signed>
+static const Type* unsigned_mod_value(PhaseGVN* phase, const Node* mod) {
+  const Type* t1 = phase->type(mod->in(1));
+  const Type* t2 = phase->type(mod->in(2));
+  if (t1 == Type::TOP) {
+    return Type::TOP;
+  }
+  if (t2 == Type::TOP) {
+    return Type::TOP;
+  }
+
+  // 0 MOD X is 0
+  if (t1 == TypeClass::ZERO) {
+    return TypeClass::ZERO;
+  }
+  // X MOD X is 0
+  if (mod->in(1) == mod->in(2)) {
```

Comment:
This is analogous to code for ModI/LNode::Ideal.

Reply:
That may be true. But `uncast` would be still a good idea, so that we can see through `Cast` nodes. I think it could then also be added to `ModI/L`.

@merykitty do you have an easy way to unit test this, i.e. to have a test with a Cast?

Reply:
@eme64 I didn't mean to imply it's not useful. But I think it would be better to do these changes to the existing code for ModI/L and DivI/L should be done in a separate RFE.

Reply:
I don't see the logic in "Doing it the same as `ModI/LNode` then changing all of them together" instead of "Doing the new thing in the better way then changing the old thing to match it"

Reply:
Yeah, it is a trade-off. I think this here is small enough to just change it now in the same RFE. But if you prefer to do it in a separate RFE, that's fine too for me.

Reply:
If I make these changes here now, I need to come up with tests for these uncasts, which will probably take me a while (I have no experience with this yet and I can't tell how long it might take me) and the PR will stall in the meantime. If I open a follow-up RFE, I can take a look at this independently and this PR, whose main intention was to take over the main applicable optimizations from the signed to unsigned can be merged already. So as @eme64 says, it's a trade-off between creating the perfect PR now or getting some good changes in and then making refinements in the same area in a later, follow-up PR.

Reply:
Drive-by comment: Such uncast optimizations are definitely non-trivial changes as they tend to trigger other issues by allowing subgraphs to be folded that would otherwise not be folded. So let's make sure we have proper tests for this. In my opinion, putting some of this work into a separate RFE is perfectly fine.

Reply:
Other comments were addressed. RFE for this one: https://bugs.openjdk.org/browse/JDK-8345170